### DISCUSSION TITLE
What exactly does the recovery_as_collision do from move_and_collide() for Jolt?

### DISCUSSION THREAD

**Discussion by Gamepro5** [godot-jolt/godot-jolt#547] [Questions] [answered]
The Godot docs say it gives more reliable collision normals for floors.

> If recovery_as_collision is true, any depenetration from the recovery phase is also reported as a collision; this is used e.g. by [CharacterBody3D](https://docs.godotengine.org/en/stable/classes/class_characterbody3d.html#class-characterbody3d) for improving floor detection during floor snapping.

From my use of it, it seems to be rather problematic because it seems include collision normals that I am not moving into for instance, if I am standing next to a wall and I am doing a test move directly down (to check for floors and do snapping), I will receive the normal of this wall. When I am on the corner of a stair step, it will sometimes report both the <0,1,0> normal of the floor and a <1,0,0> normal of the wall that 90 degree angle makes.

Turning it off introduces other bugs in my character controller, so I would really like to know exactly what it does and if there are any workaround for not getting a collision normal that you aren't directly casting into.

I can try to make MRPs to show exactly what I mean if I'm not explaining it well (I also found a situation where even with it off I would get a collision normal for something that is behind me and in the exact opposite direction of my movement vector)

**Comment by mihe** [comment, marked as answer]
As mentioned [here](https://github.com/godot-jolt/godot-jolt/discussions/526#discussioncomment-6614828), `move_and_collide` is broken up into three phases. The first being depenetration, the second being the cast/move and the third being the actual collision check. If `recovery_as_collision` is off, then the third phase will only happen if the second phase actually hit something during the cast. If `recovery_as_collision` is on, then the third phase will also happen if your character/body had to depenetrate from something (with the safe margin considered) during the first phase.

As long as you don't venture too deep into [this `test_body_motion` method](https://github.com/godot-jolt/godot-jolt/blob/7fd05b293351c312ca74c094666c60f5eb3124e6/src/spaces/jolt_physics_direct_space_state_3d.cpp#L531) then the code itself shouldn't be too hard to parse. That's the only place where `recovery_as_collision` is considered.

**Reply by Gamepro5** [reply]
"safe margin" as in the parameter for move_and_collide() or the safe margin of the collision shape?

**Reply by mihe** [reply]
"Safe margin" typically refers to the `margin` parameter for the shape queries, like `move_and_collide`, `move_and_slide`, `cast_motion`, etc., mostly because that's what the property is called for `CharacterBody3D`.

The margin for the shapes themselves is typically referred to as just the "shape margin".

Somewhat unfortunate/annoying naming, I'll admit.
